Code to fix laser durations

jianyoulin opened the pull request:
Please review the code.

jianyoulin committed:
Jian you's first commit to fix laser durations

jianyoulin committed:
Added the code to fix laser durations

narendramukherjee commented on fix_laser_durations.py:
```diff
@@ -0,0 +1,44 @@
+#How to change value in hf5 File due to Laser sampling error (high sampling rate)
```
You want to leave a more explanatory comment up top - I would say something like "Since the digital inputs are being sampled at 30kHz, sometimes the laser durations (or onsets) are recorded for a few ms more or less than the intended length of the pulse. We scale the length of the pulse by factors of 10 - so sometimes a 2500ms pulse can become 2510 or 2490ms. This gives errors in later steps."

narendramukherjee commented on fix_laser_durations.py:
```diff
@@ -0,0 +1,44 @@
+
+trains_dig_in = hf5.list_nodes('/spike_trains')
+
+# Ask the user for the correct laser duration and convert to integers
```
I see that you are trying to change durations here, but you need to change the onset lags as well because they can have similar sampling errors as the durations. We need to have code to deal with the onset lags too.

narendramukherjee commented on fix_laser_durations.py:
```diff
@@ -0,0 +1,44 @@
+
+# Checking the laser-duration array to find sampling errors and correct them
+for i in range(len(trains_dig_in)):    
+    exec('items = np.logical_and((trains_dig_in[{}].laser_durations[:] != 0.0), (trains_dig_in[{}].laser_durations[:] != duration))'.format(i,i))
```
It is unnecessary to use exec here. I would write it as:
```
items = np.logical_and((trains_dig_in[i].laser_durations[:] != 0.0), (trains_dig_in[i].laser_durations[:] != duration))
```

Also, up top, you need to specify that this code assumes that there's only 1 unique duration in the dataset (apart from 0.0 of course)

narendramukherjee commented on fix_laser_durations.py:
```diff
@@ -0,0 +1,44 @@
+        places = np.where(items)[0]
+        exec('trains_dig_in[{}].laser_durations[places] = duration'.format(i))
+
+items = np.logical_and((trains_dig_in[3].laser_durations[:] != 0.0), (trains_dig_in[0].laser_durations[:] != duration))
```
Why is this run separately for trains_dig_in[3] ? Why is the loop above not able to do this already?

narendramukherjee reviewed (commented):
There's a couple of things that need to be changed necessarily, and another couple I am confused about. Have a look and let me know!

jianyoulin committed:
Second commit to the code

narendramukherjee commented on fix_laser_sampling _errors.py:
```diff
@@ -0,0 +1,75 @@
+hf5.close()
+
+
+#==============================================================================
```
This was some test code below right, the stuff that is hashed out? Can you delete it?

narendramukherjee commented on fix_laser_sampling _errors.py:
```diff
@@ -0,0 +1,75 @@
+durations = easygui.multenterbox(msg = 'What are the laser durations used in this experiment?', fields = ["Duration" for num in range(num_durations)])
+for i in range(len(durations)):
+    durations[i] = int(durations[i])
+    durations.append(0)
```
The for loop should look like this:
```
for i in range(len(durations)):
   durations[i] = int(durations[i])
durations.append(0)
```
In your version, you will be appending 0 to the durations list everytime the for loop runs. This works right now because you only have 1 duration - but if you have more durations, then multiple zeros will get appended.

narendramukherjee commented on fix_laser_sampling _errors.py:
```diff
@@ -0,0 +1,75 @@
+num_durations = easygui.multenterbox(msg = 'how many laser durations were used in this experiment?', fields = ['Number of Laser durations'])
+num_durations = int(num_durations[0])
+
+durations = easygui.multenterbox(msg = 'What are the laser durations used in this experiment?', fields = ["Duration" for num in range(num_durations)])
```
Two changes here:
1. You should say in the multenterbox message that you are adding the 0 duration condition automatically (so that people don't choose 2 durations, thinking that 0 is one of them).
2. Change the line slightly to this:
```
durations = easygui.multenterbox(msg = 'What are the laser durations used in this experiment?', fields = ["Duration{}".format(num + 1) for num in range(num_durations)])
```
The box just looks nicer then :P Not very important, but still!

narendramukherjee commented on fix_laser_sampling _errors.py:
```diff
@@ -0,0 +1,75 @@
+latencies = easygui.multenterbox(msg = 'What are the latencies of laser onset used for this experiment?', fields = ["Latency" for num in range(num_latencies)])
+for i in range(len(latencies)):
+    latencies[i] = int(latencies[i])
+    latencies.append(0)
```
Look at the comment for the durations step above - this line should be outside the for loop. Else, you are adding multiple zeros to the latencies list.

As I say that, I realize that having multiple zeros in the durations or latencies list is actually not a problem with the way you've implemented stuff below (which, I must say, is amazingly well done! :D). But still, having only one copy of the zero is neater and the right way to do stuff.

narendramukherjee commented on fix_laser_sampling _errors.py:
```diff
@@ -0,0 +1,75 @@
+num_latencies = easygui.multenterbox(msg = 'how many latencies were used in this experiment?', fields = ['Number of Laser onset latencies'])
+num_latencies = int(num_latencies[0])
+
+latencies = easygui.multenterbox(msg = 'What are the latencies of laser onset used for this experiment?', fields = ["Latency" for num in range(num_latencies)])
```
Make the same changes that I suggested for the durations multenterbox above.

narendramukherjee reviewed (commented):
Amazing work, Jian-You! This is really, really good, we can merge it as soon as you can push the changes I suggested here.

As I requested last time, could you write a couple more lines right at the start of the code explaining the purpose of the code? Look at my comments last time for suggestions..

jianyoulin committed:
Third commit by Jian-You

Changes:
1) add explanatory comments
2) move append(0) out of for loop
3) change labels for multenterbox message which are nicer now

jianyoulin committed:
Jian-You Commit 4

1) Add "What are the laser durations used in this experiment (control duration [i.e., 0 ms] added automatically)?"
2)change single quote to double quote
